### PR TITLE
[NET Attribute Conversion][generator] Generate NET style attributes

### PR DESCRIPTION
**Pull request by chamons**
🤞 the final large piece of the saga that is https://github.com/xamarin/xamarin-macios/issues/10170

We won't know until landing this and testing the result but it's the last _known_ hole.....

## Summary

This PR teaches our code generator to generate .NET 6 style availability attributes, adds a 4th Cecil test to verify our generated attributes, and a metric ton of API changes to satisfy that test. 

## Details

The generator work is the core of this PR, and includes:

- Hacking out chunks of generator.cs that "helpfully" remove duplicate attributes, which are no longer duplicate in the new order that NET6 attributes force upon us. See changes in `FilterMinimumVersion` and `PrintPlatformAttributes` 
- Prevent a crash when the generator processes availability attributes with no version included (example: introduced on iOS but no version). See `Is64BitiOSOnly`.
- The meat, `GetPlatformAttributesToPrint`, which synthesizes many attributes "out of thing air" from:
   - The parent context
   - Implied introduced just because the class exists on a given framework at all
   - Implied Catalyst because iOS exists
- A few cludgy hacks `PrintPlatformAttributesNoDuplicates` and `GenerateProperty` because the existing PrintPlatformAttributes did not pass down parent context down, and the refactor was dangerous/too time consuming given time pressure.

There are two intended API changes introduced by the reviews in this PR:

- `GetCurrentInputDevice` was obviously intended by availability attributes to exist on Catalyst but due to define confusion was excluded. It is an addition in `Microsoft.MacCatalyst.dll` only.
- The `NEAppRule` constructors were mis-marked on platforms, and were showing up incorrectly on Mac/Catalyst. I corrected the Catalyst one unconditionally, as we have not shipped Catalyst yet, but Mac is only fixed in NET6.

## Known API changes

image

image

## Limitations

I could spend another week or two polishing this, at minimum. It is being landed in this state due to time pressure. 

### There are bugs in the generated attributes:
- **Major**: Protocols with one set of attributes [TV (9,0)] inlined in types that differ [NoTV] currently unconditionally use the attributes from the protocol, which is wrong. This has always been wrong, but never noticed given how few of them we generated in legacy due to the de-duplication logic and lack of tests. This was deemed too difficult to fix in time since the calls to GetPlatformAttributesToPrint did not pass the "target" type, and I'd have to do wider surgery to wire it through. 
- **Major**: Obsolete attributes prevent the "implied introduced" attribute from being created, causing some enums to not generate correct attributes.
- **Minor**: Protocols that are conditionally included with defines don't have the right attributes. Example: if we include NSItemProviderReading only on iOS those inlined protocols don't have [NoTV] equivalents. 
- **Minor**: Attributes that are conditionally set as Abstract also don't set the right attributes.

Each of these has cases hit that are Ignored in test code (see `IgnoreElementsThatDoNotExistInThatAssembly`).

### Other things to note (and forgive) include:
- The gross `#if` hacks in generator.cs, which are signs that my hacked in code did not gel well with the existing logic. I should in theory extend each check to better handle the NET6 availability attribute reality and then remove logic from GetPlatformAttributesToPrint.
- GetPlatformAttributesToPrint is a god-function that does too much. It is well documented and calls into child functions where possible, but I could do better with more time.
- There are zero tests outside of the integration Cecil tests.

Despite this, given time pressure and test coverage I think it is ready to go in.

I've filed https://github.com/xamarin/xamarin-macios/issues/14802 to track the cleanup work from this PR.

### CODE REVIEW

File: src/AudioUnit/AUEnums.cs [hunk 1]
```diff
@@ -189,12 +189,20 @@ public enum AudioObjectPropertySelector : uint
 		TranslateUIDToBox = 1969841250, // 'uidb'
 		ClockDeviceList = 1668049699, //'clk#'
 		TranslateUidToClockDevice = 1969841251, // 'uidc',
+#if XAMCORE_3_0
```
Comment (rolfbjarne) [inline src/AudioUnit/AUEnums.cs:192]: The defines in this enum are somewhat confusing.

I think this would simplify it:

```cs
[NoTV][NoWatch]
#if NET
[NoiOS]
#endif
public enum AudioObjectPropertySelector : uint
{
...
#if !NET
 		[Deprecated (PlatformName.iOS, 15,0, message : "Use the 'ProcessIsMain' element instead.")]
#endif
 		[Deprecated (PlatformName.MacCatalyst, 15,0, message : "Use the 'ProcessIsMain' element instead.")]
 		[Deprecated (PlatformName.MacOSX, 12,0, message : "Use the 'ProcessIsMain' element instead.")]
 		[Obsolete ("Use the 'ProcessIsMain' element instead.")]
 		ProcessIsMaster = 1835103092, // 'mast'
 		[MacCatalyst (15,0), Mac (12,0)]
 		ProcessIsMain = 1835100526, // 'main'
...
 		[MacCatalyst (15,0), Mac (12,0)]
 		ProcessMute = 1634758765, // 'appm'
}
```

That said, the changes here look correct, so this could go in another PR.

File: src/contactsui.cs [hunk 2]
```diff
@@ -74,6 +77,7 @@ interface CNContactPickerDelegate
 		void DidClose (CNContactPicker picker);
 	}
 #else
+	[NoMac]
```
Comment (rolfbjarne) [inline src/contactsui.cs:80]: Any reason you didn't merge the `CNContactPickerDelegate` definitions like you did for `CNContactViewController`?

File: tests/cecil-tests/AttributeTest.cs [hunk 3]
```diff
@@ -153,8 +464,7 @@ void CheckAllPlatformsOnParent (ICustomAttributeProvider item, string fullName,
 		// }
 		//
 		// When run against mac, this fails as Extension does not include a mac supported of any kind attribute
-		// XXX - This test is disabled until generator work is complete, as there are too many false positives
-		// [TestCaseSource (typeof (Helper), "NetPlatformAssemblies")]
+		[TestCaseSource (typeof (Helper), "NetPlatformAssemblies")]
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:467]:
```suggestion
		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
```

File: tests/cecil-tests/AttributeTest.cs [hunk 4]
```diff
@@ -163,36 +473,31 @@ public void AllAttributedItemsMustIncludeCurrentPlatform (string assemblyPath)
 				return;
 			}
 
-#if DEBUG
-			Console.WriteLine(assemblyPath);
-#endif
+			if (!assemblyPath.Contains("Xamarin.Mac.dll")) {
```
Comment (rolfbjarne) [inline tests/cecil-tests/AttributeTest.cs:476]: There's no "Xamarin.Mac.dll" anymore for .NET (it's "Microsoft.macOS.dll"), so this seems wrong.

File: src/generator.cs [hunk 5]
```diff
@@ -3393,6 +3398,306 @@ bool FilterMinimumVersion (AvailabilityBaseAttribute aa)
...
+				return meth.DeclaringType.Namespace;
+			default:
+				throw new NotImplementedException ($"FindNamespace on {item} of type {item.GetType()}");
+		}
+	}
+
+	IEnumerable<AvailabilityBaseAttribute> AddUnlistedAvailability (MemberInfo containingClass, List<AvailabilityBaseAttribute> availability)
+	{
+		// If there are literally no attributes for a platform on a type, for non-catalyst platforms
+		// add a minimum supported introdced since it was "unlisted"
```
Comment (rolfbjarne) [inline src/generator.cs:3505]:
```suggestion
		// add a minimum supported introduced since it was "unlisted"
```

File: src/generator.cs [hunk 6]
```diff
@@ -3393,6 +3398,306 @@ bool FilterMinimumVersion (AvailabilityBaseAttribute aa)
...
+			default:
+				throw new NotImplementedException ($"ConvertKindToMatchKind with unknown kind {kind}");
+		}
+	}
+
+	// This assumes the compiler implements property methods as get_ or set_ prefixes
+	static PropertyInfo GetProperyFromGetSetMethod (MethodInfo method)
+	{
+		string name = method.Name;
+		if (name.StartsWith ("get_") || name.StartsWith ("set_")) {
```
Comment (rolfbjarne) [inline src/generator.cs:3601]:
```suggestion
		if (name.StartsWith ("get_", StringComparison.Ordinal) || name.StartsWith ("set_", StringComparison.Ordinal)) {
```

File: src/generator.cs [hunk 7]
```diff
@@ -3393,6 +3398,306 @@ bool FilterMinimumVersion (AvailabilityBaseAttribute aa)
...
+				return containingProperty;
+			}
+		}
+		if (mi is TypeInfo) {
+			return mi;
+		}
+		return FindContainingContext (mi.DeclaringType);
+	}
+
+	// We need to collect all of the availabilty attriutes walking up the chain of context.
```
Comment (rolfbjarne) [inline src/generator.cs:3624]:
```suggestion
	// We need to collect all of the availability attributes walking up the chain of context.
```

**Review by rolfbjarne — APPROVED** [review]
This looks good, just a few minor issues.

File: tests/cecil-tests/AttributeTest.cs (hunk 4 above)
Comment (chamons) [reply tests/cecil-tests/AttributeTest.cs:476]: This was a cludge to only run one test more quickly. 🤦 

Let me remove and re-run

**Comment by chamons** [comment]
@rolfbjarne - Requesting re-review just because the last commit was large (fixing the issue you found).

File: tests/cecil-tests/AttributeTest.cs (hunk 4 above)
Comment (chamons) [reply tests/cecil-tests/AttributeTest.cs]: Fixed!

File: src/contactsui.cs (hunk 2 above)
Comment (chamons) [reply src/contactsui.cs:80]: This was just found later and I started running into more time pressure to get this done.

File: src/AudioUnit/AUEnums.cs (hunk 1 above)
Comment (chamons) [reply src/AudioUnit/AUEnums.cs:192]: Like it, let's save it for a follow up PR.

File: src/generator.cs [hunk 8]
```diff
@@ -3393,6 +3399,313 @@ bool FilterMinimumVersion (AvailabilityBaseAttribute aa)
...
+			parentAvailability.AddRange(AttributeManager.GetCustomAttributes<AvailabilityBaseAttribute> (context));
+			var parentContext = FindContainingContext (context);
+			if (context == parentContext) {
+				return parentAvailability;
+			}
+			context = parentContext;
+		}
+	}
+
+	AvailabilityBaseAttribute [] GetPlatformAttributesToPrint (MemberInfo mi, Type type, MemberInfo inlinedType)
```
Comment (dalexsoto) [inline src/generator.cs:3650]: 👍 great explanation!

**Review by dalexsoto — APPROVED** [review]
LGTM 👍

**Review by rolfbjarne — APPROVED** [review]

File: tests/cecil-tests/AttributeTest.cs [hunk 9]
```diff
@@ -107,35 +124,346 @@ public void DoubleAttributedElements (string assemblyPath)
...
+			}
+
+			Assert.That (attributesWereCompiledOut, Is.Empty, $"{attributesWereCompiledOut.Count} issues found");
+			Assert.That (doesNotExistWhereClaimed, Is.Empty, $"{doesNotExistWhereClaimed.Count} issues found");
+		}
+
+		static bool IgnoreElementsThatDoNotExistInThatAssembly (string member)
+		{
+			// Xkit has many platform specific bits in AppKit/UIKit and is a mess to get right
+			if (member.StartsWith ("Kit")) {
```
[the remark this answers is not in the document]
Comment (chamons) [reply tests/cecil-tests/AttributeTest.cs:261]: Nope, we strip it in the test here:

https://github.com/xamarin/xamarin-macios/blob/06cf323805c587d3f63e5ba0d515e60db30a13b3/tests/cecil-tests/AttributeTest.cs#L446-L452

**Review by mandel-macaque — APPROVED** [review]

**Comment by chamons** [comment]
I will be reviewing the test failures locally.

**Comment by chamons** [comment]
I've filed https://github.com/xamarin/xamarin-macios/issues/14802 to track the cleanup work from this PR.

**Comment by chamons** [comment]
Unfortunately I had to add a few more commits to fix the introspection tests. I'll look at the rest tomorrow.

File: src/generator.cs [hunk 10]
```diff
@@ -3393,6 +3399,320 @@ bool FilterMinimumVersion (AvailabilityBaseAttribute aa)
...
+
+			// Now expand the implied catalyst from [iOS] a second time
+			// This is needed in some cases where the only iOS information is in the
+			// parent context, but we want to let any local iOS override a catalyst on the parent
+			AddImpliedCatalyst (memberAvailability);
+
+			// If all of this implication gives us something silly, like being introduced
+			// on a type that is on a namespace we don't support, ignore those Supported
+			StripIntroducedOnNamespaceNotIncluded (memberAvailability, context);
+			if (inlinedType != null) {
```
Comment (tj-devel709) [inline src/generator.cs:3707]: nit: just a few places in this method that could use "is not null" instead of "!= null"
but very small thing in PR of this magnitude!

**Review by tj-devel709 — APPROVED** [review]

**Comment by chamons** [comment]
Still a few test failures locally, and one generator change needed to fix some of them.

**Comment by mandel-macaque** [comment]
@chamons  I merged with main to get the api diff back

File: src/generator.cs [hunk 11]
```diff
@@ -3511,38 +3509,43 @@ bool IsInSupportedFramework (MemberInfo klass, PlatformName platform)
...
+				availability.Add (CreateNoVersionSupportedAttribute (platform));
 			}
 		}
 	}
 
 	static void AddImpliedCatalyst (List<AvailabilityBaseAttribute> memberAvailability)
 	{
-		foreach (var attr in memberAvailability.Where (v => v.Platform == PlatformName.iOS && v is not ObsoletedAttribute).ToList()) {
-			if (!memberAvailability.Any (v => (v.Platform == PlatformName.MacCatalyst) && ImpliedKindsMatch (v.AvailabilityKind, attr.AvailabilityKind))) {
+		if (!PlatformMarkedUnavailabile (PlatformName.MacCatalyst, memberAvailability) &&
```
Comment (tj-devel709) [inline src/generator.cs:3528]: PlatformMarkedUnavailabile -> PlatformMarked**Unavailable**

File: src/generator.cs [hunk 12]
```diff
@@ -3511,38 +3509,43 @@ bool IsInSupportedFramework (MemberInfo klass, PlatformName platform)
...
-		// If there are literally no attributes for a platform on a type, for non-catalyst platforms
-		// add a minimum supported introduced since it was "unlisted"
+		// If there are no unavailable attributes for a platform on a type
+		// add a supported introduced (without version) since it was "unlisted" (for non-catalyst platforms)
 		foreach (var platform in new [] { PlatformName.iOS, PlatformName.TvOS, PlatformName.MacOSX }) {
-			if (!availability.Any (v => v.Platform == platform)) {
-				if (IsInSupportedFramework (containingClass, platform)) {
-					yield return CreateMinSupportedAttribute (platform);
-				}
+			if (!PlatformMarkedUnavailabile (platform, availability) && IsInSupportedFramework (containingClass, platform)) {
```
Comment (tj-devel709) [inline src/generator.cs:3520]: PlatformMarked**Unavailable**

File: src/generator.cs [hunk 13]
```diff
@@ -3511,38 +3509,43 @@ bool IsInSupportedFramework (MemberInfo klass, PlatformName platform)
...
 			}
 		}
 	}
 
+	static bool PlatformHasIntroduced (PlatformName platform, List<AvailabilityBaseAttribute> memberAvailability) 
+	{
+		return memberAvailability.Any (v => (v.Platform == platform && v is IntroducedAttribute));
+	}
+
+	static bool PlatformMarkedUnavailabile (PlatformName platform, List<AvailabilityBaseAttribute> memberAvailability)
```
Comment (tj-devel709) [inline src/generator.cs:3544]: PlatformMarked**Unavailable**

File: src/generator.cs [hunk 14]
```diff
@@ -3684,17 +3693,17 @@ List<AvailabilityBaseAttribute> GetAllParentAttributes (MemberInfo context)
 			availabilityToConsider = availabilityToConsider.Where (x => x.Platform != PlatformName.WatchOS).ToList();
 
 			// Add any implied non-catalyst introduced (Catalyst will come later)
-			availabilityToConsider.AddRange (AddUnlistedAvailability (context, availabilityToConsider));
+			AddUnlistedAvailability (context, availabilityToConsider);
 
-			// Copy down any unavailable from the parent before expanding, since a [NoMacCatalyst] on the type trumps [iOS] on a member
-			CopyValidAttributes (memberAvailability, availabilityToConsider.Where (attr => attr.AvailabilityKind != AvailabilityKind.Introduced).Select (attr => CloneFromOtherPlatform (attr, attr.Platform)));
+			// Copy down any unavailable from the parent before expanding, since a [NoMacCatalyst	] on the type trumps [iOS] on a member
```
Comment (tj-devel709) [inline src/generator.cs:3698]: Nit: Space here on purpose?

**Review by tj-devel709 — COMMENTED** [review]
Small typo I believe in a method name, but the other new things look good to me!

File: src/generator-attributes.cs [hunk 15]
```diff
@@ -931,19 +925,61 @@ public override string ToString ()
 				builder.AppendLine ("#if __MACOS__");
 				break;
 			case PlatformName.MacCatalyst:
-				builder.AppendLine ("#if __MACCATALYST__");
+				builder.AppendLine ("#if __MACCATALYST__ && !__IOS__ && !__MACOS__");
```
Comment (rolfbjarne) [inline src/generator-attributes.cs:928]: The `&& !__MACOS__` part shouldn't be needed, but it doesn't hurt either.

**Review by rolfbjarne — APPROVED** [review]

File: src/generator-attributes.cs (hunk 15 above)
Comment (chamons) [reply src/generator-attributes.cs]: Fixed!

File: src/generator.cs (hunk 14 above)
Comment (chamons) [reply src/generator.cs]: Fixed!

File: src/generator.cs (hunk 13 above)
Comment (chamons) [reply src/generator.cs]: Fixed!

File: src/generator.cs (hunk 11 above)
Comment (chamons) [reply src/generator.cs]: Fixed!

**Comment by chamons** [comment]
Generator test failure looks legit somehow, fixing.

**Comment by chamons** [comment]
Looks like the generator test actually caught a logic bug in the generator change. In NET6 I assume that all namespaces must be one of the Apple binding one, else we must be unsupported and thus skip attributes. 

This completely ignored 3rd party bindings.

File: src/generator.cs [hunk 16]
```diff
@@ -3393,37 +3399,364 @@ bool FilterMinimumVersion (AvailabilityBaseAttribute aa)
...
+		case PlatformName.TvOS:
+			return new UnavailableAttribute(platform);
+		case PlatformName.WatchOS:
+			throw new InvalidOperationException ("CreateUnsupportedAttribute for WatchOS never makes sense");
+		default:
+			throw new NotImplementedException ();
+		}
+	}
+
+	HashSet<string> GetFrameworkListForPlatform (PlatformName platform)
```
Comment (stephen-hawley) [inline src/generator.cs:3470]: For the future, this type of code can be simplified with a switch expression:
```
var frameworkList = platform switch {
    PlatformName.iOS => Frameworks.iosframeworks,
    PlatformName.TvOS => Frameworks.tvosframeworks,
    PlatformName.MacOSX => Frameworks.macosframeworks,
    PlatformName.MacCatalyst => Frameworks.maccatalstframeworks,
    _ => new HashSet<string>(),
};
return new HashSet<string>(frameworkList.Select(x => x.ToLower (CultureInfo.InvariantCulture)));
```

File: src/generator.cs [hunk 17]
```diff
@@ -3393,37 +3399,364 @@ bool FilterMinimumVersion (AvailabilityBaseAttribute aa)
...
+			case TypeInfo type:
+				return type.Namespace;
+			case PropertyInfo prop:
+				return prop.DeclaringType.Namespace;
+			case MethodInfo meth:
+				return meth.DeclaringType.Namespace;
+			default:
+				throw new NotImplementedException ($"FindNamespace on {item} of type {item.GetType()}");
+		}
+	}
```
Comment (stephen-hawley) [inline src/generator.cs:3506]: same here:
```
static string FindNamespace (MemberInfo item) => item switch {
    TypeInfo type => type.Namespace,
    PropertyInfo prop => prop.DeclaringType.Namespace,
    MethodInfo meth => meth.DeclaringType.Namespace,
   _ => throw new NotImplementedException ("..."),
}
```
Also - should this have a case for events?

**Review by stephen-hawley — APPROVED** [review]

File: src/Metal/MTLEnums.cs [hunk 18]
```diff
@@ -832,7 +832,7 @@ public enum MTLFeatureSet : ulong {
 	[iOS (9,0)][Mac (10,11)]
 	[Native]
 	public enum MTLLanguageVersion : ulong {
-		[NoMac]
+		[NoMac][NoMacCatalyst]
```
[the remark this answers is not in the document]
Comment (chamons) [reply src/Metal/MTLEnums.cs:835]: No. We are absurdly inconsistent.

File: src/generator-attributes.cs [hunk 19]
```diff
@@ -931,19 +925,61 @@ public override string ToString ()
 				builder.AppendLine ("#if __MACOS__");
 				break;
 			case PlatformName.MacCatalyst:
-				builder.AppendLine ("#if __MACCATALYST__");
+				builder.AppendLine ("#if __MACCATALYST__ && !__IOS__");
```
[the remark this answers is not in the document]
Comment (chamons) [reply src/generator-attributes.cs:928]: Pretty much, though invariants are tough to pin down in the generator since it's older code with few tests.

File: src/generator.cs (hunk 17 above)
Comment (chamons) [reply src/generator.cs:3506]: Events are never (?) in bindings, I'd rather crash and find out than add it and hope it works.

File: src/avfoundation.cs [hunk 20]
```diff
@@ -7596,17 +7596,17 @@ interface AVFragmentedMovie_AVFragmentedMovieTrackInspection
 		AVFragmentedMovieTrack[] GetTracks (AVMediaCharacteristics mediaCharacteristic);
 
 		[Async]
-		[Watch (8,0), TV (15,0), Mac (12,0), iOS (15,0), MacCatalyst (15,0)]
```
[the remark this answers is not in the document]
Comment (chamons) [reply src/avfoundation.cs:7599]: The type itself is marked with NoTV, and my tests screamed about the inconsistency.

**Comment by chamons** [comment]
Generator test _should_ be fixed.

File: tests/cecil-tests/AttributeTest.cs [hunk 21]
```diff
@@ -107,35 +124,359 @@ public void DoubleAttributedElements (string assemblyPath)
...
```
Comment (mandel-macaque) [inline tests/cecil-tests/AttributeTest.cs:261]: All this can be a lot nicer with a switch and since we have the new c# we could put all in the same switch.

**Review by mandel-macaque — APPROVED** [review]
We might want to re-think the 10000 ones switches in the future, is getting out of hand.

**Comment by chamons** [comment]
@mandel-macaque - I fully the switches in AttributesTest.cs are absurd, but they should be temporary until https://github.com/xamarin/xamarin-macios/issues/14802 is fixed.

File: src/generator-attributes.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (stephen-hawley) [reply src/generator-attributes.cs:928]: OK - for the future:
```
static class StringBuilderExtensions {
    public static StringBuilder AppendLineIfNeeded (this StringBuilder sb)
    {
        if (sb.Length == 0) return sb;
        return sb [sb.Length - 1] == '\n' ? sb : sb.Append ('\n');
    }
    public static StringBuilder AppendLineIfNeeded (this StringBuilder sb, string text)
        => return sb.AppendLineIfNeeded().Append(text);
}
